File: src/main/java/com/crud_javatokotiln/service/ProductService.java
package com.crud_javatokotiln.service;


import com.crud_javatokotiln.dto.ProductDto;
import com.crud_javatokotiln.entity.Product;
import com.crud_javatokotiln.exception.NotFoundException;
import com.crud_javatokotiln.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    public List<Product> findAll() {
        return productRepository.findAll();
    }

    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Product not found with id: " + id));
    }

    @Transactional
    public Product create(ProductDto productDto) {
        Product product = new Product(productDto.getName(), productDto.getPrice());
        return productRepository.save(product);
    }

    @Transactional
    public Product update(Long id, ProductDto productDto) {
        Product product = findById(id);
        product.setName(productDto.getName());
        product.setPrice(productDto.getPrice());
        return productRepository.save(product);
    }

    @Transactional
    public void delete(Long id) {
        productRepository.deleteById(id);
    }
}

File: src/main/java/com/crud_javatokotiln/service/ProductService.kt
package com.crud_javatokotiln.service

import com.crud_javatokotiln.dto.ProductDto
import com.crud_javatokotiln.entity.Product
import com.crud_javatokotiln.exception.NotFoundException
import com.crud_javatokotiln.repository.ProductRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
open class ProductService(
    private val productRepository: ProductRepository
) {

    fun findAll(): List<Product> = productRepository.findAll()

    fun findById(id: Long): Product =
        productRepository.findById(id)
            .orElseThrow { NotFoundException("Product not found with id: $id") }

    @Transactional
    open fun create(productDto: ProductDto): Product {
        val name = productDto.name ?: throw IllegalArgumentException("Name cannot be null")
        val price = productDto.price ?: throw IllegalArgumentException("Price cannot be null")
        val product = Product(name, price)
        return productRepository.save(product)
    }

    @Transactional
    open fun update(id: Long, productDto: ProductDto): Product {
        val product = findById(id)
        product.name = productDto.name ?: product.name
        product.price = productDto.price ?: product.price
        return productRepository.save(product)
    }

    @Transactional
    open fun deleteById(id: Long) {
        if (!productRepository.existsById(id)) {
            throw NotFoundException("Product not found with id: $id")
        }
        productRepository.deleteById(id)
    }
}
